tools/sg-cksum/cksum.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)cksum.1	8.2 (Berkeley) 4/28/95
.\"	$Id: cksum.1,v 1.1 2006-07-27 23:37:04 damion Exp $
.\"
.Dd April 28, 1995
.Dt CKSUM 1
.Os BSD 4.4
.Sh NAME
.Nm cksum
.Nd display file checksums and block counts
.Sh SYNOPSIS
.Nm cksum
.Oo 
.Fl o Ar \&1 Xo
.No \&|
.Ar \&2
.No \&|
.Ar \&3
.Xc
.Oc
.Op Ar file ...
.Nm sum
.Op Ar file ...
.Sh DESCRIPTION
The
.Nm cksum
utility writes to the standard output three whitespace separated
fields for each input file.
These fields are a checksum
.Tn CRC ,
the total number of octets in the file and the file name.
If no file name is specified, the standard input is used and no file name
is written.
.Pp
The
.Nm sum
utility is identical to the
.Nm cksum
utility, except that it defaults to using historic algorithm 1, as
described below.
It is provided for compatibility only.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl o
Use historic algorithms instead of the (superior) default one.
.Pp
Algorithm 1 is the algorithm used by historic
.Bx
systems as the 
.Xr sum 1
algorithm and by historic
.At V
systems as the
.Xr sum
algorithm when using the
.Fl r
option.
This is a 16-bit checksum, with a right rotation before each addition;
overflow is discarded.
.Pp
Algorithm 2 is the algorithm used by historic
.At V
systems as the
default
.Xr sum
algorithm.
This is a 32-bit checksum, and is defined as follows:
.Bd -unfilled -offset indent
s = sum of all bytes;
r = s % 2^16 + (s % 2^32) / 2^16;
cksum = (r % 2^16) + r / 2^16;
.Ed
.Pp
Algorithm 3 is what is commonly called the
.Ql 32bit CRC
algorithm.
This is a 32-bit checksum.
.Pp
Both algorithm 1 and 2 write to the standard output the same fields as
the default algorithm except that the size of the file in bytes is
replaced with the size of the file in blocks.
For historic reasons, the block size is 1024 for algorithm 1 and 512
for algorithm 2.
Partial blocks are rounded up.
.El
.Pp
The default
.Tn CRC
used is based on the polynomial used for
.Tn CRC
error checking
in the networking standard
.St -iso8802-3
The
.Tn CRC
checksum encoding is defined by the generating polynomial:
.Pp
.Bd -unfilled -offset indent
G(x) = x^32 + x^26 + x^23 + x^22 + x^16 + x^12 +
     x^11 + x^10 + x^8 + x^7 + x^5 + x^4 + x^2 + x + 1
.Ed
.Pp
Mathematically, the
.Tn CRC
value corresponding to a given file is defined by
the following procedure:
.Bd -filled -offset indent
The
.Ar n
bits to be evaluated are considered to be the coefficients of a mod 2
polynomial M(x) of degree
.Ar n Ns \-1 .
These
.Ar n
bits are the bits from the file, with the most significant bit being the most
significant bit of the first octet of the file and the last bit being the least
significant bit of the last octet, padded with zero bits (if necessary) to
achieve an integral number of octets, followed by one or more octets
representing the length of the file as a binary value, least significant octet
first.
The smallest number of octets capable of representing this integer are used.
.Pp
M(x) is multiplied by x^32 (i.e., shifted left 32 bits) and divided by
G(x) using mod 2 division, producing a remainder R(x) of degree <= 31.
.Pp
The coefficients of R(x) are considered to be a 32-bit sequence.
.Pp
The bit sequence is complemented and the result is the CRC.
.Ed
.Pp
The
.Nm cksum
and
.Nm sum
utilities exit 0 on success, and >0 if an error occurs.
.Sh SEE ALSO
.Xr md5 1
.Rs
The default calculation is identical to that given in pseudo-code
in the following
.Tn ACM
article.
.Rs
.%T "Computation of Cyclic Redundancy Checks Via Table Lookup"
.%A Dilip V. Sarwate
.%J "Communications of the \\*(tNACM\\*(sP"
.%D "August 1988"
.Re
.Sh STANDARDS
The
.Nm cksum
utility is expected to conform to
.St -p1003.2-92 .
.Sh HISTORY
The
.Nm cksum
utility appeared in
.Bx 4.4 .


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/cksum/cksum.c
/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * James W. Williams of NASA Goddard Space Flight Center.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1991, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)cksum.c	8.2 (Berkeley) 4/28/95";
#endif
static const char rcsid[] =
	"$Id: cksum.c,v 1.9 1998/03/10 05:03:31 jb Exp $";
#endif /* not lint */

#include <sys/cdefs.h>
#include <sys/types.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include "extern.h"

#ifndef warn
#define warn(format, arg...) fprintf(stderr, format, ##arg)
#endif

static void usage __P((void));

int binary = 0;

int
main(argc, argv)
	int argc;
	char **argv;
{
	register int ch, fd, rval;
	u_int32_t len, val;
	char *fn;
	unsigned char nbuf[4];
	int (*cfncn) __P((int, u_int32_t *, u_int32_t *));
	void (*pfncn) __P((char *, u_int32_t, u_int32_t));

	cfncn = crc;
	pfncn = pcrc;

	while ((ch = getopt(argc, argv, "bo:")) != -1)
		switch (ch) {
		case 'o':
			if (!strcmp(optarg, "1")) {
				cfncn = csum1;
					pfncn = psum1;
			} else if (!strcmp(optarg, "2")) {
				cfncn = csum2;
				pfncn = psum2;
			} else if (*optarg == '3') {
				cfncn = crc32;
				pfncn = pcrc;
			} else {
				fprintf(stderr, "illegal argument to -o option\n");
				usage();
			}
			break;
		case 'b':
			binary++;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	fd = STDIN_FILENO;
	fn = NULL;
	rval = 0;
	do {
		if (*argv) {
			fn = *argv++;
			if ((fd = open(fn, O_RDONLY, 0)) < 0) {
				warn("%s", fn);
				rval = 1;
				continue;
			}
		}
		if (cfncn(fd, &val, &len)) {
			warn("%s", fn ? fn : "stdin");
			rval = 1;
		} else {
			if (binary) {
				nbuf[0] = (val >> 24) & 0xff;
				nbuf[1] = (val >> 16) & 0xff;
				nbuf[2] = (val >> 8) & 0xff;
				nbuf[3] = val & 0xff;
				write(1,  &nbuf[0], 4);
			} else {
				pfncn(fn, val, len);
			}
		}
		(void)close(fd);
	} while (*argv);
	exit(rval);
}

static void
usage()
{
	(void)fprintf(stderr, "usage: cksum [-b] [-o 1 | 2] [file ...]\n");
	(void)fprintf(stderr, "       sum [file ...]\n");
	exit(1);
}
